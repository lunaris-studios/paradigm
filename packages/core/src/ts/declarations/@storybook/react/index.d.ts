import * as Abstract from "@paradigmjs/abstract";
import * as Addons from "@storybook/addons";
import * as SR from "@storybook/react";

import "@storybook/react";

declare module "@storybook/react" {
	/**
	 * Metadata to configure the stories for an abstract component.
	 */
	declare type StorybookComponent =
		| React.ComponentType
		| Abstract.AbstractComponent
		| Abstract.AbstractPureComponent;
	export declare type Meta<
		TComponent extends StorybookComponent = any,
		TArgs = Addons.DefaultArgs
	> = Addons.BaseMeta<TComponent> &
		Addons.Annotations<Component.Spinner, React.ReactElement<unknown>>;
}
