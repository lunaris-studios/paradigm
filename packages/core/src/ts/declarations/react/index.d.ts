import * as SC from "styled-components";

declare module "react" {
	interface Attributes {
		css?: SC.CSSProp;
	}
}
