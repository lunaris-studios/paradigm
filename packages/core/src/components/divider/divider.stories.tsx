import * as React from "react";
import * as SR from "@storybook/react";

import * as Component from "./divider";

const Template: SR.Story<Component.IDividerProps> = (args) => (
	<Component.Divider {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};

export default {
	title: "Divider",
	component: Component.Divider,
} as SR.Meta;
