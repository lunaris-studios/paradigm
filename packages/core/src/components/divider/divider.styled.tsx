import * as Protocol from "@paradigmjs/protocol";
import * as SC from "styled-components";

// re-import `styled-components` development mode DOM classnames.
import styled from "styled-components";

/**
 * Table of Contents
 *
 * [Divider]
 */

/**
 * [Divider]
 * - [Divider.Element(ELEMENT)]
 */

interface Divider {
	Element: SC.StyledComponent<
		"div",
		any,
		IDividerElementAttrs,
		keyof IDividerElementAttrs
	>;
}

export const Divider = {} as Divider;

/**
 * [Divider.Element]
 */

interface IDividerElementProps extends SC.ThemeProps<SC.DefaultTheme> {
	/**
	 * Should be a [[solid color]], changes in color scheme
	 * are dynamic handled via relative opacity changes.
	 * @default Color.BLACK_0
	 */
	color: Protocol.Color;
}

interface IDividerElementAttrs extends IDividerElementProps {}

const DIVIDER_ELEMENT_OPACITY = Protocol.bind("scheme", {
	[Protocol.Scheme.DARK]: 0.75,
	[Protocol.Scheme.LIGHT]: 0.5,
});

const defaultDividerElementProps = Object.freeze<IDividerElementProps>({
	color: Protocol.Color.BLACK_0,
	// TODO (sam): Fix parent extension of <SC.ThemeProps>
	theme: {} as Protocol.Theme,
});

Divider.Element = styled("div").attrs(
	(props: IDividerElementProps = defaultDividerElementProps): IDividerElementAttrs => ({
		...props,
	}),
)`
	border-color: ${(props) => props.color};
	margin: ${Protocol.Snippets.px(Protocol.Space.ONE)};
	opacity: ${DIVIDER_ELEMENT_OPACITY};

	border-bottom-width: 1px;
	border-bottom-style: solid;

	/**
   * Since the element is empty, it will occupy minimal space and only show
   * the appropriate border based on direction of container.
   */
	border-right-width: 1px;
	border-right-style: solid;
`;
