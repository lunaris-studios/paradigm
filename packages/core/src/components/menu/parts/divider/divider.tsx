import classNames from "classnames";
import * as React from "react";

import * as Common from "~/common";
import * as Components from "~/components";

export interface IMenuDividerProps extends Common.IProps {
	/** This component does not support children. */
	children?: never;

	/** Optional header title. */
	title?: React.ReactNode;
}

export class MenuDivider extends React.Component<IMenuDividerProps, {}> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.MenuDivider`;

	public render() {
		const { title } = this.props;

		if (title == null) {
			// simple divider
			return <li />;
		} else {
			// section header with title
			return (
				<li>
					<Components.H6>{title}</Components.H6>
				</li>
			);
		}
	}
}
