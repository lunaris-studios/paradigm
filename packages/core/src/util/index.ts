export * from "./_can-use-dom";
export * from "./_compose";
export * from "./_function-utils";
export * from "./_is-key-board-click";
export * from "./_is-node-env";
export * from "./_js-utils";
export * from "./_pipe";
export * from "./_react-utils";
export * from "./_remove-non-html-props";
export * from "./_test-util";

import clamp from "lodash/clamp";

export { clamp };
export { default as isNull } from "lodash/isNull";
export { default as camelCase } from "lodash/camelCase";
