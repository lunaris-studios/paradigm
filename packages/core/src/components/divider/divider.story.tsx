import * as React from "react";
import * as Knobs from "@storybook/addon-knobs";
import * as Actions from "@storybook/addon-actions";

import { Divider } from "./divider";

export default {
	title: "Divider",
	component: Divider,
	decorators: [Knobs.withKnobs],
};

export const interactive = () => {
	return <Divider />;
};
interactive.story = {
	parameters: {
		jest: ["divider.test.tsx"],
	},
};

export const Vertical = () => {
	return (
		<div style={{ display: "flex", justifyContent: "column" }}>
			<div />
			<Divider />
			<div />
		</div>
	);
};

export const Horizontal = () => {
	return (
		<div style={{ display: "flex", justifyContent: "row" }}>
			<div />
			<Divider />
			<div />
		</div>
	);
};
