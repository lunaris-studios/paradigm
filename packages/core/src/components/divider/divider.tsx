import * as React from "react";

import * as Common from "~/common";
import * as Components from "~/components";

import * as Styled from "./divider.styled";

export interface IDividerProps extends Common.IProps, React.HTMLAttributes<HTMLElement> {
	/**
	 * HTML tag to use for element.
	 * @default "div"
	 */
	tagName?: keyof JSX.IntrinsicElements;
}

const defaultProps = Object.freeze<IDividerProps>({
	tagName: "div",
});

export class Divider extends Components.AbstractPureComponent<IDividerProps> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.Divider`;

	static readonly defaultProps: IDividerProps = defaultProps;

	public render() {
		const { tagName } = this.props;
		return <Styled.Divider.Element as={tagName} />;
	}
}
