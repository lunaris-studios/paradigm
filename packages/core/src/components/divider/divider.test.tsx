import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";

import * as Util from "~/util";

import * as Component from "./divider";
import * as Styled from "./divider.styled";

// @ts-ignore
import { find } from "styled-components/test-utils";

describe("Divider", () => {
	it("renders correctly", () => {
		const node = Util.wrapper(<Component.Divider />);
		const DividerContainer = find(node.baseElement, Styled.Divider.Element);
		expect(DividerContainer).toMatchSnapshot();
	});

	it("<Styled.Divider.Element> responds to color scheme change when [Scheme.DARK] is active", () => {
		const node = Util.wrapper(<Component.Divider />, { scheme: Protocol.Scheme.DARK });
		const DividerElement = find(node.baseElement, Styled.Divider.Element);
		expect(DividerElement).toHaveStyleRule("opacity", "0.75");
	});

	it("<Styled.Divider.Element> responds to color scheme change when [Scheme.LIGHT] is active", () => {
		const node = Util.wrapper(<Component.Divider />, { scheme: Protocol.Scheme.LIGHT });
		const DividerElement = find(node.baseElement, Styled.Divider.Element);
		expect(DividerElement).toHaveStyleRule("opacity", "0.5");
	});
});
