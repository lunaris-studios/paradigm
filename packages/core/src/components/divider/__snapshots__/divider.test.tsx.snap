// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider renders correctly 1`] = `
.c0 {
  margin: 4px;
  opacity: 0.75;
  border-bottom-width: 1px;
  border-bottom-style: solid;
  border-right-width: 1px;
  border-right-style: solid;
}

<div
  class="c0"
/>
`;
